docs/using-est-with-certd.md

```diff
@@ -0,0 +1,169 @@
+
+# Example with the EST path segment `test-est-cert` in the URL.
+# The path will depend on server configuration.
+url = "https://est.example.com/.well-known/est/test-est-cert"
+
+# This example will use the EST bootstrap certificate and key to authenticate,
+# but a cert and key ID for the identity certificate must still be provided.
+# These credentials will automatically be created, so it is sufficient to provide
+# a set of unique IDs that will not be used for any other credentials.
+identity_cert = "test-est-cert-identity-id"
+identity_pk = "test-est-cert-identity-id"
+
+# The credentials to use upon initial authentication with the EST server.
+# After that, certd will automatically create, use, and renew identity_cert and identity_pk.
+bootstrap_identity_cert = "test-est-bootstrap-cert-id"
+bootstrap_identity_pk = "test-est-bootstrap-key-id"
+
+# Load the bootstrap certificate from a file into certd.
+# This file must be readable by aziotcs.
+[preloaded_certs]
+test-est-bootstrap-cert-id = "file:///path/to/file.pem"
+
+# Allow the specified user to create this certificate.
+# Replace `1000` with the relevant user ID. Note that the root user has access to all certificates.
+[[principal]]
+uid = 1000
+certs = ["test-est-cert"]
+```
+
+The keyd configuration file authorizes aziotcs to access the required keys.
+
+`/etc/aziot/keyd/config.d/test-est-cert.toml`
+
+```toml
+# Load the bootstrap certificate key from a file into key.
+# This file must be readable by aziotks.
+[preloaded_keys]
+test-est-bootstrap-key-id = "file:///path/to/file.pem"
+
+# Authorize aziotcs to access the necessary keys.
+[[principal]]
+# Replace with output of `id -u aziotcs`
+uid = 997
+
+# Should contain the key IDs of bootstrap_identity_pk and identity_pk.
+keys = ["test-est-bootstrap-key-id", "test-est-cert-identity-id"]
+```
+
+## Testing EST on the command line
+
+After configuring Certificates Service, you can test certificate issuance over EST from the command line. Generate a CSR and make a request to Certificates Service to create a new certificate.
+
+```sh
+# Generate a key for CSR.
+openssl genrsa -out key.pem
+
+# Generate CSR. Note that the common name must be provided either here in the CSR
+# or in certd's configuration.
+#
+# This command generates a simple CSR that doesn't have attributes or other useful
+# information. For a production certificate, you will likely need to add attributes
+# or configure the EST server to automatically issue certificates with the required
+# attributes.
+openssl req -new -key key.pem -subj "/CN=test-est-cert" -out req.pem
+
+# Escape newlines in CSR.
+csr=$(awk '{printf "%s\\n", $0}' req.pem)
+
+# Make the request to Certificates Service.
+# The user making this request must be root or in the aziotcs group.
+curl --unix-socket /run/aziot/certd.sock http://localhost/certificates?api-version=2020-09-01 \
+    -H "content-type: application/json" \
+    --data "{\"certId\": \"test-est-cert\", \"csr\": \"$csr\"}"
```

Comment:
Use `jq --arg` to construct the request body, which does the escaping for you.

Reply:
Fixed

Reply:
Just do it inline instead of saving to a file, eg https://github.com/Azure/iot-identity-service/blob/f6294bdd027284a27a03d83b4c5dc736d98e0a32/ci/iothub-get-twin.sh#L247-L256

Reply:
Fixed